README.pod
=pod

=encoding UTF-8

=head1 NAME

Math::Shape::Vector - A 2d vector library in cartesian space

=head1 VERSION

version 0.1

=head1 SYNOPSIS

    use Math::Shape::Vector;

    my $v1 = Math::Shape::Vector->new(3, 5);
    my $v2 = Math::Shape::Vector->new(1, 17);

    $v1->add_vector($v2);
    $v1->negate;
    $v1->multiply(5);
    $v1->is_equal($v2);

=head1 DESCRIPTION

This module contains 2d vector-based objects intended as base classes for 2d games programming. Most of the objects have collision detection (among other methods). The objects available are:

=over

=item *

L<Math::Shape::Vector> - a 2d vector (this module)

=item *

L<Math::Shape::Line> - an infinite 2d line

=item *

L<Math::Shape::LineSegment> - a finite 2d line (with a start and end)

=item *

L<Math::Shape::Range> - a number range (e.g 1 through 20)

=item *

L<Math::Shape::Circle> - a 2d Circle

=item *

L<Math::Shape::Rectangle> - a 2d axis-oriented rectangle

=item *

L<Math::Shape::OrientedRectangle> - a 2d oriented rectangle

=back

=for HTML <a href="https://travis-ci.org/sillymoose/Math-Shape-Vector"><img src="https://travis-ci.org/sillymoose/Math-Shape-Vector.svg?branch=master"></a> <a href='https://coveralls.io/r/sillymoose/Math-Shape-Vector'><img src='https://coveralls.io/repos/sillymoose/Math-Shape-Vector/badge.png' alt='Coverage Status' /></a>

=head1 METHODS

=head2 new

Create a new vector. Requires two numerical arguments for the origin and magnitude.

    my $vector = Math::Shape::Vector->new(3, 5);

=head2 add_vector

Adds a vector to the vector object, returning a new vector object with the resulting x & y values.

    my $new_vector = $vector->add_vector($vector_2);

=head2 subtract_vector

Subtracts a vector from the vector object, returning a new vector object with the resulting x & y values.

    my $new_vector = $vector->subtract_vector($vector_2);

=head2 negate

Returns a new vector with negated values values e.g. (1,3) becomes (-1, -3).

    my $new_vector = $vector->negate();

=head2 is_equal

Compares a vector to the vector object, returning 1 if they are the same or 0 if they are different.

    $vector->is_equal($vector_2);

=head2 multiply

Returns a new vector object with the x and y values multiplied by a number.

    my $new_vector = $vector->multiply(3);

=head2 divide

Returns a new vector object with the x and y values divided by a number.

    my $new_vector = $vector->divide(2);

=head2 rotate

Returns a new vector with the x and y values rotated in radians.

    use Math::Trig ':pi';

    my $new_vector = $vector->rotate(pi);

=head2 rotate_90

Returns a new vector object with the x and y values rotated 90 degrees anti-clockwise.

    my $new_vector = $vector->rotate_90;

=head2 dot_product

Returns the dot product. Requires another Math::Shape::Vector object as an argument.

=head2 length

Returns the vector length.

    my $length = $vector->length;

=head2 convert_to_unit_vector

Returns a new vector object with a length of 1.

    my $unit_vector = $vector->convert_to_unit_vector;

=head2 project

Maps the vector to another vector, returning a new vector object. Requires a Math::Shape::Vector object as an argument.

    my $new_vector = $vector->project($vector_2);

=head2 is_parallel

Boolean method that returns 1 if the vector is parallel with another vector else returns zero. Requires a Math::Shape::Vector object as an argument.

    my $v2 = Math::Shape::Vector(1, 2);

    if ($v->is_parallel($v2))
    {
        # do something
    }

=head2 enclosed_angle

Returns the enclosed angle of another vector. Requires a Math::Shape::Vector object as an argument.

    my $v2 = Math::Shape::Vector(4, 2);
    my $enclosed_angle = $v->enclosed_angle($v2);

=head2 collides

Boolean method that returns 1 if the vector collides with another L<Math::Shape::Vector> library object or not or 0 if not. Requires a Math::Shape::Vectorlibrary object as an argument

    my $v1 = Math::Shape::Vector(4, 2);
    my $v2 = Math::Shape::Vector(4, 2);

    $v1->collides($v2); # 1

    my $circle = Math::Shape::Circle->new(0, 0, 3); # x, y and radius
    $v1->collides($circle); # 0

=head1 REPOSITORY

L<https://github.com/sillymoose/Math-Shape-Vector.git>

=head1 THANKS

The source code for this object was inspired by the code in Thomas Schwarzl's 2d collision detection book L<http://www.collisiondetection2d.net>.

=head1 AUTHOR

David Farrell <sillymoos@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by David Farrell.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Change history for Math::Shape::Vector

0.09    2014/09/23
        -Added circle-vector, circle-line collision detection

0.08    2014/09/21
        -Added Circle class
        -Added Rectangle class
        -Added OrientedRectangle class

0.07    2014/09/21
        -Removed test dependency on Math::Shape::Point

0.06    2014/09/20
        -Added LineSegment collides() and Line one_one_side methods

0.05    2014/09/13
        -Added rotate_90(), enclosed_angle(), collides()
        -Added Math::Shape::Utils
        -Added Math::Shape::Line
        -Added Math::Shape::LineSegment
        -Renamed get_length() and get_dot_product(), removing 'get_'

0.04    2014/09/02
        -Added get_dot_product(), project(), rotate(), convert_to_unit_vector()

0.03    2014/09/01
        -Added get_length()

0.02    2014/09/01
        -Updated POD

0.01    2014/09/01
        -Initial build


lib/Math/Shape/Line.pm
use strict;
use warnings;
package Math::Shape::Line;

use 5.008;
use Carp;
use Math::Shape::Vector;

# ABSTRACT: a 2d vector-based infinite line

=head1 METHODS

=head2 new

Constructor, requires 4 values: the x,y values for the base and direction vectors.

    my $line = Math::Shape::Line->new(1, 2, 3, 4);

=cut

sub new {
    croak 'incorrect number of args' unless @_ == 5;
    my ($class, $x1, $y1, $x2, $y2) = @_;
    bless { base        => Math::Shape::Vector->new($x1, $y1),
            direction   => Math::Shape::Vector->new($x2, $y2),
          }, $class;
}

=head2 is_equivalent

Boolean method returns 1 if the line is equivalent to another line object. Lines are equivalent when they are parallel and the base vector of one line occurs along the other line. Requires a Math::Shape::Line object as an argument.

    if ($line->is_equivalent($other_line)
    {
        ...
    }

=cut

sub is_equivalent
{
    croak 'must pass a line object' unless $_[1]->isa('Math::Shape::Line');
    unless( $_[0]->{direction}->is_parallel($_[1]->{direction}) )
    {
        0;
    }
    else
    {
        my $base = $_[0]->{base}->subtract_vector($_[1]->{base});
        $base->is_parallel($_[0]->{direction});
    }
}

=head2 one_one_side

Boolean method that returns 1 if both points of a LineSegment object are on the same side of the line. Requires a L<Math::Shape::LineSegment> object as an argument.

=cut

sub on_one_side
{
    croak 'project not called with argument of type Math::Shape::Line' unless $_[1]->isa('Math::Shape::LineSegment');
    my ($self, $segment) = @_;

    my $vector_d1 = $segment->{start}->subtract_vector($self->{base});
    my $vector_d2 = $segment->{end}->subtract_vector($self->{base});
    my $vector_n = $self->{direction}->rotate_90;

    $vector_n->dot_product($vector_d1) * $vector_n->dot_product($vector_d2) 
        > 0 ? 1 : 0;
}


=head2 collides

Boolean method that returns 1 if the line collides with another L<Math::Shape::Vector> library object or not or 0 if not. Requires a Math::Shape::Vectorlibrary object as an argument

    my $l1 = Math::Shape::Line(4, 2);
    my $l2 = Math::Shape::Line(4, 2);

    $l1->collides($l2); # 1

=cut

sub collides
{
    my ($self, $other_obj) = @_;

    if ($other_obj->isa('Math::Shape::Line'))
    {
        if($self->{direction}->is_parallel($other_obj->{direction}))
        {
            $self->is_equivalent($other_obj);
        }
        else
        {
            1;
        }
    }
    elsif ($other_obj->isa('Math::Shape::LineSegment'))
    {
        $self->on_one_side($other_obj) ? 0 : 1;
    }
    elsif ($other_obj->isa('Math::Shape::OrientedRectangle'))
    {
        $other_obj->collides($self);
    }
    elsif ($other_obj->isa('Math::Shape::Circle'))
    {
        $other_obj->collides($self);
    }
    elsif ($other_obj->isa('Math::Shape::Rectangle'))
    {
        $other_obj->collides($self);
    }
    elsif ($other_obj->isa('Math::Shape::Vector'))
    {
        $other_obj->collides($self);
    }
    else
    {
        croak 'collides must be called with a Math::Shape::Vector library object';
    }
}



1;


lib/Math/Shape/Range.pm
use strict;
use warnings;
package Math::Shape::Range;

use 5.008;
use Carp;
use Math::Shape::Utils;
use List::Util qw/min max/;

# ABSTRACT: a range object which has min and max values

=head1 METHODS

=head2 new

Constructor, requires 2 values: minimum and maximum floating point numbers.

    my $range = Math::Shape::Range->new(3.5, 4);

=cut

sub new
{
    croak 'incorrect number of args' unless @_ == 3;
    my ($class, $min, $max) = @_;
    bless { min => $min, max => $max }, $class;
}

=head2 sort

Returns a new range object with the min and max attributes sorted.

    my $sorted_range = $range->sort;

=cut

sub sort
{
    my $self = shift;

    if ($self->{min} > $self->{max})
    {
        Math::Shape::Range->new(
            $self->{max},
            $self->{min},
        );
    }
    else {
        $self;
    }
}

=head2 is_overlapping

Boolean method which returns 1 if the range object overlaps with another range, or 0 if not. Requires another range object as an argument.

    my $range1 = Math::Shape::Range->new(1, 10);
    my $range2 = Math::Shape::Range->new(11, 20);
    my $range3 = Math::Shape::Range->new(15, 25);
    $range1->is_overlapping($range2); # 0
    $range2->is_overlapping($range3); # 1

=cut

sub is_overlapping
{
    croak 'Must provide another Math::Shape::Range object as argument' unless $_[1]->isa('Math::Shape::Range');
    overlap($_[0]->{min}, $_[0]->{max}, $_[1]->{min}, $_[1]->{max});
}

=head2 hull

Returns a new Range object of the hull of two ranges. Requires another Math::Shape::Range object as an argument.

=cut

sub hull
{
    croak 'Must provide another Math::Shape::Range object as argument' unless $_[1]->isa('Math::Shape::Range');
    my ($self, $other_range) = @_;

    my $min = min $self->{min}, $other_range->{min};
    my $max = max $self->{max}, $other_range->{max};

    Math::Shape::Range->new($min, $max);
}


1;


lib/Math/Shape/Utils.pm
use strict;
use warnings;
package Math::Shape::Utils;

use Math::Trig ':pi';
use Carp 'croak';

# ABSTRACT: Utility methods used by the Math::Shape::Vector module

BEGIN {
    require Exporter;
    use base qw(Exporter);
    our @EXPORT = qw(degrees_to_radians radians_to_degrees overlap equal_floats minimum maximum clamp_on_range);
    our @EXPORT_OK = ();
}

sub degrees_to_radians
{
    $_[0] * pi / 180.0;
}

sub radians_to_degrees
{
    $_[0] / pi * 180.0;
}

sub overlap
{
    my ($minA, $maxA, $minB, $maxB) = @_;
    $minB <= $maxA && $minA <= $maxB ? 1 : 0;
}

sub equal_floats
{
    my $threshold = 1.0 / 8192.0;
    abs($_[0] - $_[1]) < $threshold ? 1 : 0;
}

sub minimum
{
    $_[0] < $_[1] ? $_[0] : $_[1];
}

sub maximum
{
    $_[0] > $_[1] ? $_[0] : $_[1];
}

sub clamp_on_range
{
    croak 'clamp_on_range() called without enough args. It requires 3 arguments: x, min & max' unless @_ == 3;
    my ($x, $min, $max) = @_;

    if ($x < $min)
    {
        $min;
    }
    elsif ($x > $max)
    {
        $max;
    }
    else
    {
        $x;
    }
}

1;


lib/Math/Shape/Circle.pm
use strict;
use warnings;
package Math::Shape::Circle;

use 5.008;
use Carp;
use Math::Shape::Vector;
use Math::Shape::Rectangle;

# ABSTRACT: a 2d circle

=head1 METHODS

=head2 new

Constructor, requires 3 values: the x,y values for the center point and a radius

    my $circle = Math::Shape::Circle->new(1, 2, 54);

=cut

sub new {
    croak 'incorrect number of args' unless @_ == 4;
    my ($class, $x, $y, $r) = @_;
    bless { center => Math::Shape::Vector->new($x, $y),
            radius => $r,
          }, $class;
}

=head2 collides

Boolean method returns 1 if circle collides with another object, else returns 0. Requires another L<Math::Shape::Vector> library object as an argument.

    my $is_collision = $circle->collides($other_circle);

=cut

sub collides
{
    my ($self, $other_obj) = @_;

    if ($other_obj->isa('Math::Shape::Circle'))
    {
        my $center = $self->{center}->subtract_vector($other_obj->{center});
        $center->length <= $self->{radius} + $other_obj->{radius} ? 1 : 0;
    }
    elsif ($other_obj->isa('Math::Shape::Vector'))
    {
        my $center = $self->{center}->subtract_vector($other_obj);
        $center->length <= $self->{radius} ? 1 : 0;
    }
    elsif ($other_obj->isa('Math::Shape::Line'))
    {
        my $center = $self->{center}->subtract_vector($other_obj->{base});
        my $project = $center->project($other_obj->{direction});

        my $base_vector = $other_obj->{base}->add_vector($project);
        $self->collides($base_vector);
    }
    elsif ($other_obj->isa('Math::Shape::LineSegment'))
    {
        # test collision of both LineSegment start/end points
        return 1 if $self->collides($other_obj->{start});
        return 1 if $self->collides($other_obj->{end});

        # test collision of nearest point on LineSegment with circle
        my $d  = $other_obj->{end}->subtract_vector($other_obj->{start});
        my $lc = $self->{center}->subtract_vector($other_obj->{start});
        my $p  = $lc->project($d);
        my $nearest = $other_obj->{start}->add_vector($p);

        $self->collides($nearest)
            && $p->length <= $d->length
            && 0 <= $p->dot_product($d)
            ? 1 : 0;
    }
    elsif ($other_obj->isa('Math::Shape::Rectangle'))
    {
        my $clamped_vector = $other_obj->clamp($self->{center});

        $self->collides($clamped_vector);
    }
    elsif ($other_obj->isa('Math::Shape::OrientedRectangle'))
    {
        # transform OrientedRectangle into Rectangle
        my $r_size = $other_obj->{half_extend}->multiply(2);
        my $lr = Math::Shape::Rectangle->new(0, 0, $r_size->{x}, $r_size->{y});

        # transform $self into local Circle in coordinates of other_obj
        my $distance = $self->{center}->subtract_vector($other_obj->{center});
        $distance = $distance->rotate(- $other_obj->{rotation});
        my $center = $distance->add_vector($other_obj->{half_extend});
        my $lc = Math::Shape::Circle->new($center->{x}, $center->{y}, $self->{radius});

        # check local objects collide
        $lc->collides($lr);
    }
    else
    {
        croak 'collides must be called with a Math::Shape::Vector library object';
    }
}
1;


lib/Math/Shape/Rectangle.pm
use strict;
use warnings;
package Math::Shape::Rectangle;

use 5.008;
use Carp;
use Math::Shape::Vector;
use Math::Shape::Utils;
use Math::Shape::Line;
use Math::Shape::LineSegment;
use Math::Shape::Range;

# ABSTRACT: an axis-aligned 2d rectangle

=head1 METHODS

=head2 new

Constructor, requires 4 values: the x,y values for the origin and the x,y values for the size.

    my $rectangle = Math::Shape::Rectangle->new(1, 1, 2, 4);
    my $width = $rectangle->{size}->{x}; # 2

=cut

sub new {
    croak 'incorrect number of args' unless @_ == 5;
    my ($class, $x, $y, $l, $h) = @_;
    bless { origin => Math::Shape::Vector->new($x, $y),
            size   => Math::Shape::Vector->new($l, $h),
          }, $class;
}

=head2 clamp

Takes a vector object and returns a new vector object whose x & y coordinates are "clamped" to the size of the rectangle. Requires a L<Math::Shape::Vector> object as an argument.

    my $clamped_vector = $rectangle->clamp($vector);

=cut

sub clamp
{
    croak 'clamp must be called with a Math::Shape::Vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $vector) = @_;

    my $clamp_x = clamp_on_range($vector->{x}, $self->{origin}->{x}, $self->{origin}->{x} + $self->{size}->{x});
    my $clamp_y = clamp_on_range($vector->{y}, $self->{origin}->{y}, $self->{origin}->{y} + $self->{size}->{y});
    Math::Shape::Vector->new($clamp_x, $clamp_y);
}

=head2 corner

Returns a L<Math::Shape::Vector> object representing a corner point of the rectangle. Requires a number between 0 and 3.

    my $corner = $rectangle->corner(2); # get the 3rd corner

=cut

sub corner
{
    croak 'Incorrect number of arguments for corner(). Requires a number between 0 and 3.' unless @_ == 2;
    my ($self, $nr) = @_;

    my $corner;
    my $mod = $nr % 4;

    if ($mod == 0)
    {
        $corner = Math::Shape::Vector->new(
            $self->{origin}{x} + $self->{size}{x},
            $self->{origin}{y},
        );
    }
    elsif ($mod == 1)
    {
        $corner = Math::Shape::Vector->new(
            $self->{origin}{x},,
            $self->{origin}{y},
        );
        $corner->add_vector($self->{size});
    }
    elsif ($mod == 2)
    {
        $corner = Math::Shape::Vector->new(
            $self->{origin}{x},
            $self->{origin}{y} + $self->{size}{y},
        );
    }
    elsif ($mod == 3)
    {
        $corner = Math::Shape::Vector->new(
            $self->{origin}{x},
            $self->{origin}{y},
        );
    }
    else
    {
        croak 'corner() not called with a number between 0 and 3';
    }
}

=head2 separating_axis

Boolean method that returns 1 if the axis is outside of the rectangle, or 0 if not. Requires a L<Math::Shape::LineSegment> object as an argument.

=cut

sub separating_axis
{
    croak 'separating_axis() requires a Math::Shape::LineSegment object as an argument' unless $_[1]->isa('Math::Shape::LineSegment');
    my ($self, $axis) = @_;

    my $n = $axis->{start}->subtract_vector($axis->{end});
    my $point0 = $self->corner(0);
    my $point1 = $self->corner(1);
    my $point2 = $self->corner(2);
    my $point3 = $self->corner(3);

    my $r_edge_a = Math::Shape::LineSegment->new(
        $point0->{x},
        $point0->{y},
        $point1->{x},
        $point1->{y},
    );
    my $r_edge_range_a = $r_edge_a->project($n);
    my $r_edge_b = Math::Shape::LineSegment->new(
        $point2->{x},
        $point2->{y},
        $point3->{x},
        $point3->{y},
    );
    my $r_edge_range_b = $r_edge_b->project($n);
    my $r_projection = $r_edge_range_a->hull($r_edge_range_b);
    my $axis_range = $axis->project($n);
    $axis_range->is_overlapping($r_projection)
        ? 0 : 1;
}

=head2 enlarge

Returns a new rectangle object increased to the size of a vector. Requires a L<Math::Shape::Vector> object as an argument.

    my $larger_rectangle = $rectangle->enlarge($vector);

=cut

sub enlarge
{
    croak 'enlarge() must be called with a Math::Shape::Vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v) = @_;

    my $size = Math::Shape::Vector->new(
        maximum($self->{origin}{x} + $self->{size}{x}, $v->{x}),
        maximum($self->{origin}{y} + $self->{size}{y}, $v->{y}),
    );

    my $origin = Math::Shape::Vector->new(
        minimum($self->{origin}{x}, $v->{x}),
        minimum($self->{origin}{y}, $v->{y}),
    );
    my $enlarged_size = $size->subtract_vector($origin);

    Math::Shape::Rectangle->new(
        $origin->{x},
        $origin->{y},
        $enlarged_size->{x},
        $enlarged_size->{y},
    );
}

=head2 collides

Boolean method returns 1 if the rectangle collides with another object, else returns 0. Requires another L<Math::Shape::Vector> library object as an argument.

    my $is_collision = $rectangle->collides($other_rectangle);

    my $circle = Math::Shape::Circle->new(54, 19, 30);
    if ($rectangle->collides($circle))
    {
        ...
    }

=cut

sub collides {
    my ($self, $other_obj) = @_;

    if ($other_obj->isa('Math::Shape::Rectangle'))
    {
        my $a_left   = $self->{origin}{x};
        my $a_right  = $a_left + $self->{size}{x};
        my $b_left   = $other_obj->{origin}->{x};
        my $b_right  = $b_left + $other_obj->{size}{x};
        my $a_bottom = $self->{origin}{y};
        my $a_top    = $a_bottom + $self->{size}{y};
        my $b_bottom = $other_obj->{origin}{y};
        my $b_top    = $b_bottom + $other_obj->{size}{y};

        # overlap returns 1 / 0 already, so no need to use ternary to force 1/0 response
        overlap($a_left, $a_right, $b_left, $b_right)
        && overlap($a_bottom, $a_top, $b_bottom, $b_top);
    }
    elsif ($other_obj->isa('Math::Shape::Vector'))
    {
        my $left    = $self->{origin}{x};
        my $right   = $left + $self->{size}{x};
        my $bottom  = $self->{origin}{y};
        my $top     = $bottom + $self->{size}{y};

        # use ternary here as Perl will return undef if false, but we need 0
        $left <= $other_obj->{x}
            && $bottom <= $other_obj->{y}
            && $other_obj->{x} <= $right
            && $other_obj->{y} <= $top
            ? 1 : 0;
    }
    elsif ($other_obj->isa('Math::Shape::Line'))
    {
        my $n = $other_obj->{direction}->rotate_90;
        my $c1 =  $self->{origin};
        my $c2 =  $c1->add_vector($self->{size});
        my $c3 =  Math::Shape::Vector->new($c2->{x}, $c1->{y});
        my $c4 =  Math::Shape::Vector->new($c1->{x}, $c2->{y});
        $c1 = $c1->subtract_vector($other_obj->{base});
        $c2 = $c2->subtract_vector($other_obj->{base});
        $c3 = $c3->subtract_vector($other_obj->{base});
        $c4 = $c4->subtract_vector($other_obj->{base});

        my $dp1 = $n->dot_product($c1);
        my $dp2 = $n->dot_product($c2);
        my $dp3 = $n->dot_product($c3);
        my $dp4 = $n->dot_product($c4);

        # use ternary here as Perl will return undef if false, but we need 0
        ($dp1 * $dp2 <= 0)
            || ($dp2 * $dp3 <= 0)
            || ($dp3 * $dp4 <= 0)
            ? 1 : 0;
    }
    elsif ($other_obj->isa('Math::Shape::LineSegment'))
    {
        # convert LineSegment into an infinite line and test for collision
        my $base = $other_obj->{start};
        my $direction = $other_obj->{end}->subtract_vector($other_obj->{start});
        my $s_line = Math::Shape::Line->new($base->{x}, $base->{y}, $direction->{x}, $direction->{y});
        return 0 unless $self->collides($s_line);

        # convert both objects to ranges and check for overlap along x axis
        my $r_range_x = Math::Shape::Range->new(
            $self->{origin}{x},
            $self->{origin}{x} + $self->{size}{x},
        );
        my $s_range_x = Math::Shape::Range->new(
            $other_obj->{start}{x},
            $other_obj->{end}{x},
        );
        $s_range_x = $s_range_x->sort;
        return 0 unless $s_range_x->is_overlapping($r_range_x);

        # convert both objects to ranges and check for overlap along y axis
        my $r_range_y = Math::Shape::Range->new(
            $self->{origin}{y},
            $self->{origin}{y} + $self->{size}{y},
        );
        my $s_range_y = Math::Shape::Range->new(
            $other_obj->{start}{y},
            $other_obj->{end}{y},
        );
        $s_range_y = $s_range_y->sort;
        return 0 unless $s_range_y->is_overlapping($r_range_y);
    }
    elsif ($other_obj->isa('Math::Shape::OrientedRectangle'))
    {
        # get rectangular hull of oriented rectangle
        # if no collision with hull, we're good
        my $or_hull = $other_obj->hull;
        return 0 unless $self->collides($or_hull);

        # if oriented rectangle edge 0 is a separating axis, we're good
        my $or_edge_0 = $other_obj->get_edge(0);
        return 0 if $self->separating_axis($or_edge_0);

        # if oriented rectangle edge 1 is a separating axis, we're good
        my $or_edge_1 = $other_obj->get_edge(1);
        return 0 if $self->separating_axis($or_edge_1);

        # must be collision
        1;
    }
    elsif ($other_obj->isa('Math::Shape::Circle'))
    {
        # if it's a circle use the circle's collision method
        $other_obj->collides($self);
    }
    else
    {
        croak 'collides must be called with a Math::Shape::Vector library object';
    }
}

1;


lib/Math/Shape/Vector.pm
use strict;
use warnings;
package Math::Shape::Vector;

use 5.008;
use Carp;
use Math::Shape::Utils;
use Math::Trig;

# ABSTRACT: A 2d vector library in cartesian space

=for HTML <a href="https://travis-ci.org/sillymoose/Math-Shape-Vector"><img src="https://travis-ci.org/sillymoose/Math-Shape-Vector.svg?branch=master"></a> <a href='https://coveralls.io/r/sillymoose/Math-Shape-Vector'><img src='https://coveralls.io/repos/sillymoose/Math-Shape-Vector/badge.png' alt='Coverage Status' /></a>

=head1 SYNOPSIS

    use Math::Shape::Vector;

    my $v1 = Math::Shape::Vector->new(3, 5);
    my $v2 = Math::Shape::Vector->new(1, 17);

    $v1->add_vector($v2);
    $v1->negate;
    $v1->multiply(5);
    $v1->is_equal($v2);


=head1 DESCRIPTION

This module contains 2d vector-based objects intended as base classes for 2d games programming. Most of the objects have collision detection (among other methods). The objects available are:

=over

=item *

L<Math::Shape::Vector> - a 2d vector (this module)

=item *

L<Math::Shape::Line> - an infinite 2d line

=item *

L<Math::Shape::LineSegment> - a finite 2d line (with a start and end)

=item *

L<Math::Shape::Range> - a number range (e.g 1 through 20)

=item *

L<Math::Shape::Circle> - a 2d Circle

=item *

L<Math::Shape::Rectangle> - a 2d axis-oriented rectangle

=item *

L<Math::Shape::OrientedRectangle> - a 2d oriented rectangle

=back

=head1 METHODS

=head2 new

Create a new vector. Requires two numerical arguments for the origin and magnitude.

    my $vector = Math::Shape::Vector->new(3, 5);

=cut

sub new {
    croak 'incorrect number of arguments' unless @_ == 3;
    return bless { x => $_[1],
                   y => $_[2] }, $_[0];
}

=head2 add_vector

Adds a vector to the vector object, returning a new vector object with the resulting x & y values.

    my $new_vector = $vector->add_vector($vector_2);

=cut

sub add_vector {
    croak 'must pass a vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v2) = @_;

    Math::Shape::Vector->new(
        $self->{x} + $v2->{x},
        $self->{y} + $v2->{y},
    );
}

=head2 subtract_vector

Subtracts a vector from the vector object, returning a new vector object with the resulting x & y values.

    my $new_vector = $vector->subtract_vector($vector_2);

=cut

sub subtract_vector {
    croak 'must pass a vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v2) = @_;

    Math::Shape::Vector->new(
        $self->{x} - $v2->{x},
        $self->{y} - $v2->{y},
    );
}

=head2 negate

Returns a new vector with negated values values e.g. (1,3) becomes (-1, -3).

    my $new_vector = $vector->negate();

=cut

sub negate {
    my $self = shift;

    Math::Shape::Vector->new(
        - $self->{x},
        - $self->{y},
    );
}

=head2 is_equal

Compares a vector to the vector object, returning 1 if they are the same or 0 if they are different.

    $vector->is_equal($vector_2);

=cut

sub is_equal {
    croak 'must pass a vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v2) = @_;
    $self->{x} == $v2->{x} && $self->{y} == $v2->{y}
        ? 1 : 0;
}

=head2 multiply

Returns a new vector object with the x and y values multiplied by a number.

    my $new_vector = $vector->multiply(3);

=cut

sub multiply {
    croak 'incorrect number of args' unless @_ == 2;
    my ($self, $multiplier) = @_;

    Math::Shape::Vector->new(
        $self->{x} * $multiplier,
        $self->{y} * $multiplier,
    );
}

=head2 divide

Returns a new vector object with the x and y values divided by a number.

    my $new_vector = $vector->divide(2);

=cut

sub divide {
    croak 'incorrect number of args' unless @_ == 2;
    my ($self, $divisor) = @_;

    # avoid division by zero
    Math::Shape::Vector->new(
        ($divisor ? $self->{x} / $divisor : 0),
        ($divisor ? $self->{y} / $divisor : 0),
    );
}

=head2 rotate

Returns a new vector with the x and y values rotated in radians.

    use Math::Trig ':pi';

    my $new_vector = $vector->rotate(pi);

=cut

sub rotate {
    croak 'incorrect number of args' unless @_ == 2;
    my ($self, $radians) = @_;

    Math::Shape::Vector->new(
        $self->{x} * cos($radians) - $self->{y} * sin($radians),
        $self->{x} * sin($radians) + $self->{y} * cos($radians),
    );
}

=head2 rotate_90

Returns a new vector object with the x and y values rotated 90 degrees anti-clockwise.

    my $new_vector = $vector->rotate_90;

=cut

sub rotate_90
{
    my $self = shift;

    Math::Shape::Vector->new(
        - $self->{y},
        $self->{x},
    );
}

=head2 dot_product

Returns the dot product. Requires another Math::Shape::Vector object as an argument.

=cut

sub dot_product {
    croak 'must pass a vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v2) = @_;
    $self->{x} * $v2->{x} + $self->{y} * $v2->{y};
}

=head2 length

Returns the vector length.

    my $length = $vector->length;

=cut

sub length {
    my $self = shift;
    # avoid division by zero for null vectors
    my $sum_of_squares = ( $self->{x} || 0 ) ** 2
                         + ( $self->{y} || 0 ) ** 2;
    return 0 unless $sum_of_squares;
    sqrt $sum_of_squares;
}

=head2 convert_to_unit_vector

Returns a new vector object with a length of 1.

    my $unit_vector = $vector->convert_to_unit_vector;

=cut

sub convert_to_unit_vector {
    my $self = shift;

    my $length = $self->length;
    $length = 1 unless $length > 0;
    $self->divide($length)
}

=head2 project

Maps the vector to another vector, returning a new vector object. Requires a Math::Shape::Vector object as an argument.

    my $new_vector = $vector->project($vector_2);

=cut

sub project {
    croak 'must pass a vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v2) = @_;

    my $d = $v2->dot_product($v2);

    if ($d > 0) {
        $v2->multiply( $self->dot_product($v2) / $d );
    }
    else {
        $v2;
    }
}

=head2 is_parallel

Boolean method that returns 1 if the vector is parallel with another vector else returns zero. Requires a Math::Shape::Vector object as an argument.

    my $v2 = Math::Shape::Vector(1, 2);

    if ($v->is_parallel($v2))
    {
        # do something
    }

=cut

sub is_parallel
{
    croak 'must pass a vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v2) = @_;
    my $vector_na = $self->rotate_90;
    equal_floats(0, $vector_na->dot_product($v2));
}

=head2 enclosed_angle

Returns the enclosed angle of another vector. Requires a Math::Shape::Vector object as an argument.

    my $v2 = Math::Shape::Vector(4, 2);
    my $enclosed_angle = $v->enclosed_angle($v2);

=cut

sub enclosed_angle
{
    croak 'must pass a vector object' unless $_[1]->isa('Math::Shape::Vector');
    my ($self, $v2) = @_;

    my $ua = $self;
    $ua->unit_vector;

    my $ub = $v2;
    $ub->unit_vector;

    acos( $ua->dot_product($ub) );
}


=head2 collides

Boolean method that returns 1 if the vector collides with another L<Math::Shape::Vector> library object or not or 0 if not. Requires a Math::Shape::Vectorlibrary object as an argument

    my $v1 = Math::Shape::Vector(4, 2);
    my $v2 = Math::Shape::Vector(4, 2);

    $v1->collides($v2); # 1

    my $circle = Math::Shape::Circle->new(0, 0, 3); # x, y and radius
    $v1->collides($circle); # 0

=cut

sub collides
{
    my ($self, $other_obj) = @_;

    if ($other_obj->isa('Math::Shape::Vector'))
    {
        $self->{x} == $other_obj->{x} && $self->{y} == $other_obj->{y} ? 1 : 0;
    }
    elsif ($other_obj->isa('Math::Shape::LineSegment'))
    {
        # test collision of nearest point on LineSegment with vector
        my $d  = $other_obj->{end}->subtract_vector($other_obj->{start});
        my $lp = $self->subtract_vector($other_obj->{start});
        my $pr  = $lp->project($d);

        $lp->is_equal($pr)
            && $pr->length <= $d->length
            && 0 <= $pr->dot_product($d)
            ? 1 : 0;

    }
    elsif ($other_obj->isa('Math::Shape::Line'))
    {
        # test if vector collides with base
        return 1 if $self->collides($other_obj->{base});

        # test if vector lies on the direction
        my $lp = $self->subtract_vector($other_obj->{base});
        $lp->is_parallel($other_obj->{direction});
    }
    elsif ($other_obj->isa('Math::Shape::OrientedRectangle'))
    {
        $other_obj->collides($self);
    }
    elsif ($other_obj->isa('Math::Shape::Circle'))
    {
        $other_obj->collides($self);
    }
    elsif ($other_obj->isa('Math::Shape::Rectangle'))
    {
        $other_obj->collides($self);
    }
    else
    {
        croak 'collides must be called with a Math::Shape::Vector library object';
    }
}

=head1 REPOSITORY

L<https://github.com/sillymoose/Math-Shape-Vector.git>

=head1 THANKS

The source code for this object was inspired by the code in Thomas Schwarzl's 2d collision detection book L<http://www.collisiondetection2d.net>.

=cut

1;
